_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-resolve-redirect.php

```diff
@@ -0,0 +1,94 @@
+<?php
+/**
+ * REST API endpoint for resolving URL redirects.
+ *
+ * @package Jetpack
+ * @since 8.0.0
+ */
+
+/**
+ * Resolve URL redirects.
+ *
+ * @since 8.0.0
+ */
+class WPCOM_REST_API_V2_Endpoint_Resolve_Redirect extends WP_REST_Controller {
+	/**
+	 * Constructor.
+	 */
+	public function __construct() {
+		$this->namespace = 'wpcom/v2';
+		$this->rest_base = 'resolve-redirect';
+		// This endpoint *does not* need to connect directly to Jetpack sites.
+		add_action( 'rest_api_init', array( $this, 'register_routes' ) );
+	}
+
+	/**
+	 * Register the route.
+	 */
+	public function register_routes() {
+		// GET /sites/<blog_id>/resolve-redirect/<url> - Follow 301/302 redirects on a URL, and return the final destination.
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base . '/(?P<url>.+)',
+			array(
+				'args'   => array(
+					'url' => array(
+						'description'       => __( 'The URL to check for redirects.', 'jetpack' ),
+						'type'              => 'string',
+						'required'          => 'true',
+						'validate_callback' => 'wp_http_validate_url',
+					),
+				),
+				array(
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => array( $this, 'follow_redirect' ),
+					'permission_callback' => 'is_user_logged_in',
```

Comment:
I went with `is_user_logged_in`, since this endpoint isn't really tied to editing posts, it's just a helper for resolving redirects. Requiring logging in is mostly about preventing abuse.